File: tutorial/T06-Condition.cpp
/***************************
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T06-Condition.cpp
@Time: 2021/6/19 10:44 下午
@Desc: 本例主要演示，根据条件进行逻辑判断的功能
***************************/

#include "MyGCondition/MyCondition.h"
#include "MyGCondition/MyParamCondition.h"
#include "MyGNode/MyNode1.h"
#include "MyGNode/MyNode2.h"
#include "MyGNode/MyReadParamNode.h"
#include "MyGNode/MyWriteParamNode.h"

using namespace CGraph;

void tutorial_condition() {
    CStatus status;
    GPipelinePtr pipeline = GPipelineFactory::create();
    GElementPtr a, b_condition, c, d_condition = nullptr;

    b_condition = pipeline->createGGroup<MyCondition>({
        pipeline->createGNode<MyNode1>(GNodeInfo("conditionNodeB0", 1)),
        pipeline->createGNode<MyNode2>(GNodeInfo("conditionNodeB1", 1)),
        pipeline->createGNode<MyNode1>(GNodeInfo("conditionNodeB2", 1))
    });    // 生成 b_condition。执行的时候，根据choose()的返回值，在B0,B1,B2中选择一个执行

    d_condition = pipeline->createGGroup<MyParamCondition>({
        pipeline->createGNode<MyNode1>(GNodeInfo("paramConditionNodeD0", 1)),
        pipeline->createGNode<MyNode1>(GNodeInfo("paramConditionNodeD1", 1)),
        pipeline->createGNode<MyNode1>(GNodeInfo("paramConditionNodeD2", 1))
    });

    status += pipeline->registerGElement<MyWriteParamNode>(&a, {}, "writeNodeA", 1);
    status += pipeline->registerGElement<MyCondition>(&b_condition, {a}, "conditionB", 1);
    status += pipeline->registerGElement<MyReadParamNode>(&c, {b_condition}, "readNodeC", 1);
    status += pipeline->registerGElement<MyParamCondition>(&d_condition, {c}, "conditionD", 1);
    if (!status.isOK()) {
        return;
    }

    status = pipeline->init();

    for (int i = 0; i < 3; i++) {
        status += pipeline->run();
        std::cout << "[CGraph] tutorial_condition, loop : " << i + 1 << ", and run status = " << status.getCode() << std::endl;
    }

    status = pipeline->destroy();

    GPipelineFactory::remove(pipeline);
}


int main() {
    tutorial_condition();
    return 0;
}

File: python/tutorial/T06-Condition.py
"""
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T06-Condition
@Time: 2025/2/25 23:34
@Desc: 
"""

from pycgraph import GPipeline

from MyGNode.MyNode1 import MyNode1
from MyGNode.MyNode2 import MyNode2
from MyGCondition.MyCondition import MyCondition
from MyGCondition.MyParamCondition import MyParamCondition
from MyGNode.MyWriteParamNode import MyWriteParamNode
from MyGNode.MyReadParamNode import MyReadParamNode


def tutorial_condition():
    b0, b1, b2 = MyNode1("conditionNodeB0"), MyNode2("conditionNodeB1"), MyNode1("conditionNodeB2")
    b_condition = MyCondition([b0, b1, b2])

    d0, d1, d2 = MyNode1("conditionNodeD0"), MyNode1("conditionNodeD1"), MyNode1("conditionNodeD2")
    d_condition = MyParamCondition([d0, d1, d2])

    pipeline = GPipeline()
    a, c = MyWriteParamNode(), MyReadParamNode()

    pipeline.registerGElement(a, set(), "writeNodeA")
    pipeline.registerGElement(b_condition, {a}, "conditionB")
    pipeline.registerGElement(c, {b_condition}, "readNodeC")
    pipeline.registerGElement(d_condition, {c}, "conditionD")

    pipeline.init()
    for i in range(0, 3):
        status = pipeline.run()
        print("==== tutorial_condition: loop : {0}, and run status = {1}".format(i + 1, status.getCode()))

    pipeline.destroy()


if __name__ == '__main__':
    tutorial_condition()
